userapi/src/main/java/fk/prof/userapi/api/ProfileStoreAPIImpl.java

```diff
@@ -150,88 +148,80 @@ public void getProfilesInTimeWindow(Future<List<AggregatedProfileNamingStrategy>
-            // save the future, so that it can be notified when the loading visit finishes
-            saveRequestedFuture(fileNameKey, future);
-            // set the timeout for this future
-            vertx.setTimer(profileLoadTimeout, timerId -> timeoutRequestedFuture(fileNameKey, future));
-
-            if (!fileLoadInProgress) {
-                workerExecutor.executeBlocking((Future<AggregatedProfileInfo> f) -> profileLoader.load(f, filename),
-                        true,
-                        result -> completeAggregatedProfileLoading(cache, result, fileNameKey));
-            }
-        } else {
-            future.complete(cachedProfileInfo);
-        }
+    public void load(Future<AggregatedProfileInfo> future, AggregatedProfileNamingStrategy filename) {
+        clusterAwareCache.getAggregatedProfile(filename).setHandler(future.completer());
     }
 
-    @Override
-    synchronized public void loadSummary(Future<AggregationWindowSummary> future, AggregatedProfileNamingStrategy filename) {
-
-        if (!filename.isSummaryFile) {
+    public void loadSummary(Future<AggregationWindowSummary> future, AggregatedProfileNamingStrategy filename) {
+        if(!filename.isSummaryFile) {
             future.fail(new IllegalArgumentException(filename.getFileName(0) + " is not a summaryFile"));
             return;
         }
 
         String fileNameKey = filename.getFileName(0);
-
         AggregationWindowSummary cachedProfileInfo = summaryCache.getIfPresent(fileNameKey);
-        if (cachedProfileInfo == null) {
-            boolean fileLoadInProgress = futuresForLoadingFiles.containsKey(fileNameKey);
-            // save the future, so that it can be notified when the loading visit finishes
-            saveRequestedFuture(fileNameKey, future);
+        if(cachedProfileInfo == null) {
+            boolean fileLoadInProgress = filesBeingLoaded.containsKey(fileNameKey);
+            saveRequestedFutureInMap(fileNameKey, future, filesBeingLoaded);
+
             // set the timeout for this future
-            vertx.setTimer(profileLoadTimeout, timerId -> timeoutRequestedFuture(fileNameKey, future));
+            vertx.setTimer(loadTimeout, timerId -> timeoutRequestedFutureInMap(fileNameKey, future, filesBeingLoaded));
 
-            if (!fileLoadInProgress) {
+            if(!fileLoadInProgress) {
                 workerExecutor.executeBlocking((Future<AggregationWindowSummary> f) -> profileLoader.loadSummary(f, filename),
-                        true,
-                        result -> completeAggregatedProfileLoading(summaryCache, result, fileNameKey));
+                    true,
+                    result -> {
+                        if(result.succeeded()) {
+                            summaryCache.put(fileNameKey, result.result());
+                        }
+                        filesBeingLoaded.remove(fileNameKey).complete(result.map(r -> r));
```

Comment:
Earlier it was there because of type issues, auto casting to the expected type. But i think it later on resolved because of some changes made to types.
Fixed it.